### DISCUSSION TITLE
Using push + migrate messed up my db. Am I using them correctly and how to fix it?

### DISCUSSION THREAD

**Discussion by peguerosdc** [prisma/prisma#25712] [Q&A]
## Context

I have my db with an existing schema and an existing set of 9 migrations already applied. I want to change my schema and create a new migration, but I need to prototype it first as I'm not sure of how the final version will look like. Hence, I need a way to:

1. Apply changes to my existing schema without having to worry about migrations.
2. When I am happy with the result, create a migration to be applied after my 9 migrations.

## Problem

I tried to do so by following [the somehow vague recommendation](https://www.prisma.io/docs/orm/prisma-migrate/workflows/prototyping-your-schema#can-i-use-prisma-migrate-and-db-push-together) from the official docs that says:

> Can I use Prisma Migrate and db push together?
> Use db push to prototype a change to an existing schema, then run prisma migrate dev to generate a migration from your changes (you will be asked to reset)

So I made changes to my `schema.prisma` and pushed them to my local db using `prisma db push`. I did this several times until I was happy with the resulting schema and then I tried to use `prisma migrate dev` to create a migration, which resulted in the following error:

```
Error: P3006

Migration `20240815003419_init` failed to apply cleanly to the shadow database.
Error:
ERROR: relation "X" already exists
   0: schema_core::state::DevDiagnostic
             at schema-engine/core/src/state.rs:268
```

That migration is the first one of my set of 9, which means that prisma is trying to apply all the set of migrations from scratch.

I tried to fix it by running `prisma db push` using a version of my schema exactly as it is after my 9th migration. So, my db now looks like as it was before I started developing my new change and after looking at the `_prisma_migrations` table I see my 9 migrations applied correctly. Then, I tried running `prisma migrate dev` but got the same error.

If I run `prisma migrate status` I get the following message, which suggests that everything is good:

```
9 migrations found in prisma/migrations

Database schema is up to date!
```

Then, I tried to fix it by running `prisma migrate reset` using my old schema and everything seems to run correctly. Again, `prisma migrate status` says everything is up to date, but if I run `prisma migrate dev` using my new schema I get the same error. 

In my last attempt, I tried to run `prisma migrate resolve --applied 20240815003419_init`, but I get an error saying that the migration is already applied (which is consistent with my `_prisma_migrations` table).

## Questions

* How can I fix my db? That is, sync it with my schema after the 9th migration and create a new migration for my new schema.
* What is the correct development flow that I should follow the next time I want to prototype a change before creating a migration?

### Expected behavior (optional)

I was expecting that `prisma db push` or `prisma migrate reset` would leave my db in a "clean" state when ran with my old schema, so that then changing the schema and running `prisma migrate dev` would create the migration that I wanted.

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: macOS
- Database: PostgreSQL
- Node.js version: v22.6.0
- Run `prisma -v` to see your Prisma version and paste it

```
prisma                  : 5.14.0
@prisma/client          : 5.14.0
Computed binaryTarget   : darwin
Operating System        : darwin
Architecture            : x64
Node.js                 : v22.6.0
Query Engine (Node-API) : libquery-engine e9771e62de70f79a5e1c604a2d7c8e2a0a874b48 (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Schema Engine           : schema-engine-cli e9771e62de70f79a5e1c604a2d7c8e2a0a874b48 (at node_modules/@prisma/engines/schema-engine-darwin)
Schema Wasm             : @prisma/prisma-schema-wasm 5.14.0-25.e9771e62de70f79a5e1c604a2d7c8e2a0a874b48
Default Engines Hash    : e9771e62de70f79a5e1c604a2d7c8e2a0a874b48
Studio                  : 0.500.0
```

**Comment by ludralph** [comment]
Hi @peguerosdc 👋 

> How can I fix my db? That is, sync it with my schema after the 9th migration and create a new migration for my new schema.

The approach to resolve this is to use baselining. You will need to:

a) First, reset your database to the state after your 9th migration.
b) Then, follow the baselining process as described in the [documentation on baselining](https://www.prisma.io/docs/orm/prisma-migrate/workflows/baselining):
```
prisma migrate diff --from-empty --to-schema-datamodel prisma/schema.prisma --script > prisma/migrations/init/migration.sql
prisma migrate resolve --applied init
```
c) After baselining, you should be able to create a new migration for your changes using:
```
prisma migrate dev --name your_new_migration_name
```

> What is the correct development flow that I should follow the next time I want to prototype a change before creating a migration?

The recommended workflow for prototyping changes before creating a migration is:

a) Use `prisma db push` to quickly prototype and iterate on your schema changes.
b) Once you're satisfied with the changes, run `prisma migrate dev` to generate a migration from your changes.
c) Prisma will prompt you to reset the database. This is expected behavior, as mentioned in the [documentation](https://www.prisma.io/docs/orm/prisma-migrate/workflows/prototyping-your-schema).

> Expected behaviour

Your expectation was partially correct. `prisma migrate reset` should indeed bring your database to a clean state based on your migration history. However, `db push` doesn't interact with the migration history, which can lead to the issues you encountered. See the [documentation](https://www.prisma.io/docs/orm/prisma-migrate/workflows/development-and-production#development-environments).

For future prototyping, consider using a separate development database for `db push` experiments. When you're ready to create a migration, switch back to your main development database and use `migrate dev`.

**Comment by peguerosdc** [comment]
Thanks for the reply! I just deleted my db and created a new one to get rid of the shadow db problem, but I have one question about this:

> The recommended workflow for prototyping changes before creating a migration is:
a) Use prisma db push to quickly prototype and iterate on your schema changes.
b) Once you're satisfied with the changes, run prisma migrate dev to generate a migration from your changes.
c) Prisma will prompt you to reset the database. This is expected behavior, as mentioned in the [documentation](https://www.prisma.io/docs/orm/prisma-migrate/workflows/prototyping-your-schema).
[...]
For future prototyping, consider using a separate development database for db push experiments. When you're ready to create a migration, switch back to your main development database and use migrate dev.

Would you say that using a separate db for prototyping is a **strong recommendation**? I ask because unless I'm missing something, the steps as you described them are what I did and I think are the same as described in ["Prototyping with an existing migration history"](https://www.prisma.io/docs/orm/prisma-migrate/workflows/prototyping-your-schema#prototyping-with-an-existing-migration-history), but step "b" failed.

I still don't know what caused the error in the first place, so it's very likely I'm missing something, but if not using a separate db poses a high risk of running into problems with the shadow db, it'd be good to have it explicitly documented somewhere :)

**Reply by ludralph** [reply]
Hi @peguerosdc 

Using a separate database for prototyping with db push is a good practice, although it's not always strictly necessary. The recommendation stems from the potential complications that can arise when mixing `db push` and `migrate dev` on the same database. it's important to note that using `db push` and then switching to `migrate dev` can sometimes lead to issues, especially when there's an existing migration history. The documentation states that when you run `migrate dev` after prototyping with `db push`, Prisma Migrate will prompt you to reset the database because the changes made manually and with `db push` are not part of the migration history.

Prisma uses a shadow database to detect schema drift and generate new migrations. The shadow database is created and managed automatically by Prisma in most cases. However, for cloud-hosted databases, you might need to configure it manually using the shadowDatabaseUrl in your Prisma schema [Shadow database documentation](https://www.prisma.io/docs/orm/prisma-migrate/understanding-prisma-migrate/shadow-database#cloud-hosted-shadow-databases-must-be-created-manually).
While using a separate database for prototyping isn't explicitly documented as a strong recommendation, it can help avoid potential conflicts between your prototyping changes and the existing migration history. This approach can reduce the risk of unexpected issues when transitioning from `db push` to `migrate dev`.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
